src/NetworkDevice.cpp

```diff
@@ -0,0 +1,28 @@
+#include "NetworkDevice.hpp"
+#include "NetworkSimulator.hpp"
+
+NetworkDevice::NetworkDevice(DeviceType type, double processing_delay_ns):
+        type(type),
+        processing_delay_ns(processing_delay_ns) { }
+
+void NetworkDevice::Enqueue(Packet p) {
+    buffer.push(p);
+}
+
+Packet NetworkDevice::Dequeue() {
+    Packet result = buffer.front();
+    buffer.pop();
+    return result;
+}
+
+bool NetworkDevice::Empty() const {
+    return buffer.empty();
+}
+
+std::shared_ptr<NetworkDevice> NetworkDevice::NextDevice() const {
```

Comment:
As we have the global routing table, I think, NextDevice amd PrevDevice are not needed anymore? The routing table can be access when needed, without the need for NextDevice and PrevDevice. Moreover, there could be multiple `next devices`.